module/validation/seal_validator.go

```diff
@@ -0,0 +1,261 @@
+		if err != nil {
+			return nil, fmt.Errorf("could not get block payload %x: %w", ancestorID, err)
+		}
+
+		// Collect execution results from receipts.
+		for _, receipt := range payload.Receipts {
+			resultID := receipt.ExecutionResult.ID()
+			// incorporatedResults should contain the _first_ appearance of the
+			// ExecutionResult. We are traversing the fork backwards, so we can
+			// overwrite any previously recorded result.
+
+			// ATTENTION:
+			// Here, IncorporatedBlockID (the first argument) should be set
+			// to ancestorID, because that is the block that contains the
+			// ExecutionResult. However, in phase 2 of the sealing roadmap,
+			// we are still using a temporary sealing logic where the
+			// IncorporatedBlockID is expected to be the result's block ID.
+			incorporatedResults[resultID] = flow.NewIncorporatedResult(
+				receipt.ExecutionResult.BlockID,
+				&receipt.ExecutionResult,
+			)
+
+		}
+
+		ancestorID = ancestor.ParentID
+	}
+
+	// Loop forward across the fork and try to create a valid chain of seals.
+	// blockIDs, as populated by the previous loop, is in reverse order.
+	for i := len(blockIDs) - 1; i >= 0; i-- {
+		// if there are no more seals left, we can exit earlier
+		if len(byBlock) == 0 {
+			return last, nil
+		}
+
+		// return an error if there are still seals to consider, but they break
+		// the chain
+		blockID := blockIDs[i]
+		seal, found := byBlock[blockID]
+		if !found {
+			return nil, engine.NewInvalidInputErrorf("chain of seals broken (missing: %x)", blockID)
+		}
+
+		delete(byBlock, blockID)
+
+		// check if we have an incorporatedResult for this seal
+		incorporatedResult, ok := incorporatedResults[seal.ResultID]
+		if !ok {
+			return nil, engine.NewInvalidInputErrorf("seal %x does not correspond to a result on this fork", seal.ID())
+		}
+
+		// check the integrity of the seal
+		err := s.validateSeal(seal, incorporatedResult.Result)
+		if err != nil {
+			if engine.IsInvalidInputError(err) {
+				return nil, fmt.Errorf("payload includes invalid seal (%x), %w", seal.ID(), err)
+			}
+
+			return nil, fmt.Errorf("unexpected seal validation error %w", err)
+		}
+
+		last = seal
+	}
+
+	// at this point no seals should be left
+	if len(byBlock) > 0 {
+		return nil, engine.NewInvalidInputErrorf("not all seals connected to state (left: %d)", len(byBlock))
+	}
+
+	return last, nil
+}
+
+// validateSeal performs integrity checks of single seal. To be valid, we
+// require that seal:
+// 1) Contains correct number of approval signatures, one aggregated sig for each chunk.
+// 2) Every aggregated signature contains valid signer ids. module.ChunkAssigner is used to perform this check.
+// 3) Every aggregated signature contains valid signatures.
+// Returns:
+// * nil - in case of success
+// * engine.InvalidInputError - in case of malformed seal
+// * exception - in case of unexpected error
+func (s *sealValidator) validateSeal(seal *flow.Seal, executionResult *flow.ExecutionResult) error {
```

Comment:
I think we need to check if the `requireOneApproval` flag is set or not.
https://github.com/onflow/flow-go/blob/4cc4a950368b35c15b2027e6d018282ff499684a/cmd/consensus/main.go#L70
This is the flag that the consensus matching engine uses to check if it actually should collect any approvals. When the flag is not set, it will seal results without any signatures. We use the flag to smooth the transition to the full sealing logic.
So in this case we could just return `true` directly if the flag is set

Reply:
> I am not sure about moving the logic that checks if a seal corresponds to a valid IncorporatedResult on the fork, out of the Mutator and into the seal-validator.
> 
> I thought the idea of the receipt and seal validators was to perform only those checks that do not depend on the rest of the chain, but only checks like verifying signatures.
> 
> This is what it looks like for guarantees and receipts at the moment, where all the loops to peruse the chain are in the `Mutator` not the particular entity validator.
> 
> I am not sure if the plan is to move things out of the Mutator, but it would be preferable to remain consistent in my opinion.

I will try to explain my motivation why I did this way. Currently to validate both seal and receipt you need to have some knowledge about chain state, there is no way around. While receipt validation could be decoupled into single receipt validation without context knowledge, it cannot be done for seals because of `IncorporatedResult`. Moreover I don't think it's a bad thing to delegate complete block validation to some other actor, it decouples logic and makes testing easier. I would actually do the same thing for `receiptExtend`

Reply:
I see what you mean. In this case it would be good to consider the code in the builder as well:
https://github.com/onflow/flow-go/blob/894d5c048b64936461c1fec741af1f3b7be03996/module/builder/consensus/builder.go#L95-L107
It has almost the same code as the mutator (the loops through the blockchain)
It would be great to put that in a single component
For guarantees, receipts, and seals. And possibly make it more efficient because there is a lot of duplicate work.